keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+kind: ValidatingAdmissionConfiguration
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  configurationCustomResourceDefinition:
+    group: rules.example.com
+    kind: LimitRule
+    ...
+      openAPIV3Schema: ... # the below configuration is validated against this schema
+  validations:
+    - rule: "self.spec.xyz == configuration.x"
+      # ...other rule related fields here...
+    ...
+---
+apiVersion: rules.example.com/v1
+kind: LimitRule
+metadata:
+  name: "validate-xyz-rule1.example.com"
+spec:
+  match:
+  - rules: ...
+    objectSelectors: …
+    namespaceSelectors ... 
+  x: 1 
+```
+
+Pros:
+
+- Aligns closely with how policy frameworks separate policy definition from
+  policy configuration
+- Scalable beyond the maximum size of a single K8s resource
+- Easy to identify/modify/track individual policies
+- More easily allows separation-of-concerns when policy is a joint effort
+  between multiple people/teams
+- Allows for updating the abstract policy definition separate from the concrete
+  policy
+
+Cons:
+
+- Significantly more complex implementation
+  - CRDs need to be created when a ValidatingAdmissionConfiguration is created
+  - Requires the API server establish watches on the CRs to accumulate all the
+    configuration information
+- Possible to implement using Option 1. A controller can construct a
+  resource for the API in Option 1 from custom resources that define this
+  option (e.g. via a controller).
+- CRD lifecycle prevents CRs from being written until the CRD is written, which
+  means that until a ValidatingAdmissionConfiguration is written AND a CRD is
+  created for it, no CRs can be written.  Applying multiple resources
+  (e.g. kubectl apply -f <lots of yaml files>) only works if the apply
+  operations are ordered and objects are verified to exist at specific steps.
+
+#### Additional Capabilities (WIP)
+
+In addition to the high level API structure proposed above, there are quite
+a few capatiblities that are important to this enhancement that we are working
+on defining. These are all work-in-progress.
+
+Rule Message:
+
+- message formatting (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?)
+- severity (e.g. "Warning")
+- status type equivalent (e.g. Forbidden can be returned from a webhook using
+  HTTP 403) (can a match rule override this? or can it be selected via a CEL
+  expression?)
+- Similar to how each match has configuration, can each match also select
+  the severity of messages?
+
+Rule scope:
+
+- Dereferencing to deeply nested fields in CEL can be burdonsome. A field
+  containing a field path would allow a CEL rule to be scope to some node in
+  the schema.
+- Also, the idea of named scopes (e.g. "ALL_CONTAINERS") would allow a single
+  rule to be applied to all container definitions in a resource (e.g. both
+  "containers" and "initContainers" in a pod)
```

Comment:
and ephemeralContainers (seemingly inconsequential nit which actually illustrates the importance of an "all containers" helper since everyone forgets them :)

Reply:
Happy to make this list longer to emphasize the point :)